crosscluster/logical: support composite types

stevendanna opened the pull request:
See individual commits.

Fixes #127289
Release note :None

stevendanna committed:
crosscluster/logical: handle composite types in primary keys

A composite type such as collated strings store data in the Value even
when it is a primary key columns.

As a result, it cannot be correctly decoded without the
value. Deletion events from the rangefeed however, do not contain the
value.

Here, we use the previous value bytes to correctly construct the
primary key columns for the delete.

Informs #127315
Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/127317)
<!-- Reviewable:end -->

stevendanna committed:
sql/randgen: improve chances of primary index add working

It's possible the original table definition didn't include any columns
that can serve as a primary index. Here, if we find we don't have any
indexable columns, we add one.

Epic: none
Release note: None

stevendanna committed:
crosscluster/logical: unskip TestUDFWithRandomTables

Epic: none
Release note: None

msbutler commented on pkg/ccl/crosscluster/logical/lww_row_processor.go:
```diff
@@ -291,7 +300,7 @@ func (srp *sqlRowProcessor) ProcessRow(
 	}
 
 	var stats batchStats
-	if row.IsDeleted() {
```
because we can no longer rely on `row.IsDeleted()`,  delete entries will appear in the DLQ as updates/inserts :( 
https://github.com/msbutler/cockroach/blob/butler-move-origin-id/pkg/ccl/crosscluster/logical/dead_letter_queue.go#L177

stevendanna replied:
Hrmmmmmm.  Well. I pushed a bad idea. But I'm kinda thinking we just take the part of this that fixes the single column row and leave the other bits for now for 24.2

msbutler replied:
Agreed, I think we can live with this known limitation for 24.2

msbutler commented:
for 24.2, could we say composite types are not supported?

stevendanna committed:
crosscluster/logical,cdcevent: allow IsDeleted override

cdcevent.Row structs constructed by LDR may have `Value` set for
rangefeed events that originally represented deletes as a work-around
for parsing composite keys correctly.

However, it is confusing for IsDeleted() to lie since we want to be
able to reliably pass cdcevent.Row structs between functions and trust
that it has the correct state.

Here we provide a way for LDR to override the IsDeleted state.

Epic: none
Release note: none

dt commented:
> for 24.2, could we say composite types are not supported?

+1

stevendanna commented:
Extracted the single column fix to here: https://github.com/cockroachdb/cockroach/pull/127542

msbutler commented on pkg/ccl/changefeedccl/cdcevent/event.go:
```diff
@@ -141,6 +141,10 @@ func (r Row) IsDeleted() bool {
 	return r.deleted
 }
```
nit in commit message: s/traffic/track it/r

stevendanna commented:
Tagged @rharding6373 just for visibility on the small cdc-relevant change. I don't think it should impact y'all.
